cloud/services/publicips/publicips.go

```diff
@@ -18,61 +18,124 @@ package publicips
+			return nil
+		}
+
+		ipSpec := network.PublicIPAddress{
+			Sku:      &network.PublicIPAddressSku{Name: network.PublicIPAddressSkuNameStandard},
+			Name:     to.StringPtr(publicIPSpec.Name),
+			Location: to.StringPtr(s.Scope.Location()),
+			PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
+				PublicIPAddressVersion:   network.IPv4,
+				PublicIPAllocationMethod: network.Static,
+				DNSSettings: &network.PublicIPAddressDNSSettings{
+					DomainNameLabel: to.StringPtr(strings.ToLower(publicIPSpec.Name)),
+					Fqdn:            to.StringPtr(publicIPSpec.DNSName),
 				},
 			},
-		)
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(s.Scope.ResourceGroup()),
+				Role:        to.StringPtr(infrav1.CommonRole),
+				Additional:  s.Scope.AdditionalTags(),
+			})),
+		}
 
-		if err != nil {
-			return errors.Wrap(err, "cannot create public IP")
+		// https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-standard-availability-zones#zone-redundant-by-default
+
+		if err := s.Client.CreateOrUpdate(
+			ctx,
+			s.Scope.ResourceGroup(),
+			publicIPSpec.Name,
+			ipSpec,
+		); err != nil {
+			return errors.Wrap(err, "cannot create public ip")
 		}
-		klog.V(2).Infof("successfully created public IP %s", ip.Name)
+
+		klog.V(2).Infof("successfully created public IP %s", publicIPSpec.Name)
 	}
 	return nil
 }
 
 // Delete deletes the public IP with the provided scope.
 func (s *Service) Delete(ctx context.Context) error {
-	for _, ip := range s.Scope.PublicIPSpecs() {
-		klog.V(2).Infof("deleting public IP %s", ip.Name)
-		err := s.Client.Delete(ctx, s.Scope.ResourceGroup(), ip.Name)
+	for _, publicIPSpec := range s.Scope.PublicIPSpecs() {
+		klog.V(2).Infof("deleting public IP %s", publicIPSpec.Name)
+		err := s.Client.Delete(ctx, s.Scope.ResourceGroup(), publicIPSpec.Name)
 		if err != nil && azure.ResourceNotFound(err) {
 			// already deleted
 			return nil
 		}
 		if err != nil {
-			return errors.Wrapf(err, "failed to delete public IP %s in resource group %s", ip.Name, s.Scope.ResourceGroup())
+			return errors.Wrapf(err, "failed to delete public IP %s in resource group %s", publicIPSpec.Name, s.Scope.ResourceGroup())
+		}
+
+		ip, err := s.Client.Get(ctx, s.Scope.ResourceGroup(), publicIPSpec.Name)
+		if err != nil {
+			if ip.IsHTTPStatus(http.StatusNotFound) {
+				return errors.Wrapf(err, "failed to find public ip '%s' for deletion, ignoring and finishing successfully", publicIPSpec.Name)
+			}
+			return errors.Wrapf(err, "failed to fetch public ip '%s' to check if we should delete", publicIPSpec.Name)
+		}
+
+		tags := converters.MapToTags(ip.Tags)
+		isManaged := tags.HasOwned(s.Scope.ClusterName())
+		if isManaged {
+			klog.V(2).Infof("found unmanaged public ip %s, will not delete", publicIPSpec.Name)
```

Comment:
is this a breaking change? Previously created IPs with CAPZ (prior to this version) won't have the ownership tags so they won't be deleted properly if the cluster gets upgraded to this version.... Might be worth calling out in release notes. Not sure if there is a way to make this backwards compatible, unfortunately.

Reply:
if the tag doesn't exist, HasOwned returns false. So it's only true if the tag was both set and equal to true.

Reply:
right. So previously created IPs will return false to isManaged() (even though they were created by CAPZ and therefore not get deleted no? Am I missing something?

Reply:
Nope I had it backwards 🙂 I guess I'll add a release note

Reply:
see release note?